x-pack/plugin/ccr/src/internalClusterTest/java/org/elasticsearch/xpack/ccr/CcrRepositoryIT.java

```diff
@@ -655,39 +656,42 @@ public void testCcrRepositoryFailsToFetchSnapshotShardSizes() throws Exception {
-                        // this assertBusy completes because the listener is added after the InternalSnapshotsInfoService
-                        // and ClusterService preserves the order of listeners.
-                        assertBusy(() -> {
-                            List<Long> sizes = indexRoutingTable.shardsWithState(ShardRoutingState.UNASSIGNED)
-                                .stream()
-                                .filter(shard -> shard.unassignedInfo().lastAllocationStatus() == AllocationStatus.FETCHING_SHARD_DATA)
-                                .sorted(Comparator.comparingInt(ShardRouting::getId))
-                                .map(shard -> snapshotsInfoService.snapshotShardSizes().getShardSize(shard))
-                                .filter(Objects::nonNull)
-                                .filter(size -> ShardRouting.UNAVAILABLE_EXPECTED_SHARD_SIZE == size)
-                                .collect(Collectors.toList());
-                            assertThat(sizes, hasSize(numberOfShards));
-                        });
-                        waitForAllShardSnapshotSizesFailures.onResponse(null);
-                    } catch (Exception e) {
-                        throw new AssertionError("Failed to retrieve all snapshot shard sizes", e);
+            final ClusterStateListener listener = new ClusterStateListener() {
+                @Override
+                public void clusterChanged(ClusterChangedEvent event) {
+                    if (RestoreInProgress.get(event.state()).isEmpty() == false && event.state().routingTable().hasIndex(followerIndex)) {
+                        try {
+                            final IndexRoutingTable indexRoutingTable = event.state().routingTable().index(followerIndex);
+                            // this assertBusy completes because the listener is added after the InternalSnapshotsInfoService
+                            // and ClusterService preserves the order of listeners.
+                            assertBusy(() -> {
+                                List<Long> sizes = indexRoutingTable.shardsWithState(ShardRoutingState.UNASSIGNED)
+                                    .stream()
+                                    .filter(shard -> shard.unassignedInfo().lastAllocationStatus() == AllocationStatus.FETCHING_SHARD_DATA)
+                                    .sorted(Comparator.comparingInt(ShardRouting::getId))
+                                    .map(shard -> snapshotsInfoService.snapshotShardSizes().getShardSize(shard))
+                                    .filter(Objects::nonNull)
+                                    .filter(size -> ShardRouting.UNAVAILABLE_EXPECTED_SHARD_SIZE == size)
+                                    .collect(Collectors.toList());
+                                assertThat(sizes, hasSize(numberOfShards));
+                            });
+                            clusterService.removeListener(this);
```

Comment:
Only this line is the real change. The others are due to using anonymous class (so that it can referenced as `this`) instead of lambda.